source/extensions/filters/listener/http_inspector/http_inspector.cc

```diff
@@ -64,47 +65,99 @@ void Filter::onRead() {
     return;
   } else if (result.rc_ < 0) {
     config_->stats().read_error_.inc();
-    done(false);
-    return;
+    return done(false);
   }
 
   parseHttpHeader(absl::string_view(reinterpret_cast<const char*>(buf_), result.rc_));
 }
 
 void Filter::parseHttpHeader(absl::string_view data) {
-  if (absl::StartsWith(data, Filter::HTTP2_CONNECTION_PREFACE)) {
+  const size_t len = std::min(data.length(), Filter::HTTP2_CONNECTION_PREFACE.length());
+  if (Filter::HTTP2_CONNECTION_PREFACE.compare(0, len, data, 0, len) == 0) {
+    if (data.length() < Filter::HTTP2_CONNECTION_PREFACE.length()) {
+      return;
+    }
     ENVOY_LOG(trace, "http inspector: http2 connection preface found");
     protocol_ = "HTTP/2";
     done(true);
   } else {
     const size_t pos = data.find_first_of("\r\n");
+    if (pos != absl::string_view::npos) {
+      const absl::string_view request_line = data.substr(0, pos);
+      const std::vector<absl::string_view> fields =
+          absl::StrSplit(request_line, absl::MaxSplits(' ', 4));
+
+      // Method SP Request-URI SP HTTP-Version
+      if (fields.size() != 3) {
+        ENVOY_LOG(trace, "http inspector: invalid http1x request line");
+        return done(false);
+      }
+
+      if (httpMethods().count(fields[0]) == 0 || httpProtocols().count(fields[2]) == 0) {
+        ENVOY_LOG(trace, "http inspector: method: {} or protocol: {} not valid", fields[0],
+                  fields[2]);
+        return done(false);
+      }
+
+      ENVOY_LOG(trace, "http inspector: method: {}, request uri: {}, protocol: {}", fields[0],
+                fields[1], fields[2]);
+
+      protocol_ = fields[2];
+      return done(true);
+    }
 
     // Cannot find \r or \n
-    if (pos == absl::string_view::npos) {
-      ENVOY_LOG(trace, "http inspector: no request line detected");
+    ENVOY_LOG(trace, "http inspector: no request line detected");
```

Comment:
I doubt how much the code below this line can improve the performance vs its complexity, do you have any idea backing it? In a real partial HTTP/1 request line this code will be run again and again, this can lower performance especially when `checkPrefix` is not efficiently implemented. i.e. a benchmark case of sending a long request line like `GET /fooo_here_should_be_8000_bytes HTTP/1.1\r\n` by sending one bytes once could be helpful to determine.

I'm leaning to avoid that in certain case the inspector could eat up much resources (i.e. easy to issue queries to increase CPU load), than fail fast in certain case.

Reply:
I think the main purpose of the code is to reduce the false negative, especially under the scenario that HTTP request arrives byte by byte. It actually degrades the performance by introducing extra prefix check. I was thinking to implement the `checkPrefix` using trie. I think we need to trade off between the false negative and the performance here. 

I will add a benchmark test for the inspector.

Reply:
I didn't run benchmark test but I think `GET /fooo_here_should_be_8000_bytes HTTP/1.1\r\n` should be fast. (It depends on how fast `absl::hash_set` count and `absl::StrSplit`). For this example, the code will run `absl::StrSplit` and `httpMethods().count("GET")` for 8000 times.

Reply:
The point is sending `GET /fooo_here_should_be_8000_bytes HTTP/1.1\r\n` one bytes at once, so you're running partial onRead 8k times until you get full request line.